pkg/services/ngalert/state/manager.go

```diff
@@ -516,30 +516,50 @@ func translateInstanceState(state ngModels.InstanceStateType) eval.State {
 	}
 }
 
-func (st *Manager) deleteStaleStatesFromCache(logger log.Logger, evaluatedAt time.Time, alertRule *ngModels.AlertRule, takeImageFn takeImageFn) []StateTransition {
-	// If we are removing two or more stale series it makes sense to share the resolved image as the alert rule is the same.
-	// TODO: We will need to change this when we support images without screenshots as each series will have a different image
-	staleStates := st.cache.deleteRuleStates(alertRule.GetKey(), func(s *State) bool {
-		return stateIsStale(evaluatedAt, s.LastEvaluationTime, alertRule.IntervalSeconds, alertRule.GetMissingSeriesEvalsToResolve())
-	})
-	resolvedStates := make([]StateTransition, 0, len(staleStates))
+// processMissingSeriesStates receives the updated state transitions
+// that we got from the alert rule, and checks the cache for any states
+// that are not in the current evaluation. The missing states are
+// for series that are no longer present in the current evaluation.
+// For each missing state, we check if it is stale, and if so, we resolve it.
+// At the end we return the missing states so that later they can be sent
+// to the alertmanager if needed.
+func (st *Manager) processMissingSeriesStates(logger log.Logger, evaluatedAt time.Time, alertRule *ngModels.AlertRule, evalTransitions []StateTransition, takeImageFn takeImageFn) ([]StateTransition, int) {
+	// Build a set of states for series we saw in this evaluation
+	present := make(map[data.Fingerprint]struct{}, len(evalTransitions))
+	for _, tr := range evalTransitions {
+		present[tr.CacheID] = struct{}{}
+	}
+
+	missingTransitions := []StateTransition{}
+	staleCount := 0
 
-	for _, s := range staleStates {
-		logger.Info("Detected stale state entry", "cacheID", s.CacheID, "state", s.State, "reason", s.StateReason)
+	for _, s := range st.cache.getStatesForRuleUID(alertRule.OrgID, alertRule.UID) {
```

Comment:
this method creates a copy of the state. I wonder if it would be better to just iterate over state within mutex. Something like
```
st.cache.forRuleState(alertRule.OrgID, alertRule.UID, func(rs *ruleState) { ... })

Reply:
Good point, changed